### PR TITLE
PWN-8342 - implement update metadata method - to update it on server side

### PR DESCRIPTION
**Pull request by kabramovp2p**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-8342

## Description of Work

PWN-8342 - implement update metadata method - to update it on server side

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/auth/gateway/api/request/UpdateMetadataRpcRequest.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import org.p2p.solanaj.utils.crypto.Base64String
+import org.p2p.wallet.auth.gateway.api.response.UpdateMetadataResponse
+import org.p2p.wallet.utils.Base58String
+
+data class UpdateMetadataRpcRequest(
+    @Transient val ethereumAddress: String,
+    @Transient val userPublicKey: Base58String,
+    @Transient val encryptedMetadata: Base64String,
+    @Transient val timestamp: String,
+    @Transient val requestSignature: String,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/gateway/api/request/UpdateMetadataRpcRequest.kt:14]: what format has signature? base64 / base58?

File: app/src/main/java/org/p2p/wallet/auth/gateway/api/response/UpdateMetadataResponse.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package org.p2p.wallet.auth.gateway.api.response
+
+import com.google.gson.annotations.SerializedName
+
+data class UpdateMetadataResponse(
+    @SerializedName("status")
+    val status: Boolean,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/gateway/api/response/UpdateMetadataResponse.kt:7]: no need for field metadata from server?

File: app/src/main/java/org/p2p/wallet/auth/gateway/repository/mapper/GatewayResult.kt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package org.p2p.wallet.auth.gateway.repository.mapper
+
+sealed interface GatewayResult {
+
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/gateway/repository/mapper/GatewayResult.kt:4]:
```suggestion
```

File: app/src/main/java/org/p2p/wallet/auth/gateway/repository/mapper/GatewayServiceCreateWalletMapper.kt [hunk 4]
```diff
@@ -118,17 +120,17 @@ class GatewayServiceCreateWalletMapper(
...
                 ethPublic = etheriumAddress.lowercase(),
-                metaTimestamp = timestamp,
+                metaTimestamp = epochUnixTimeSeconds,
                 deviceShareDeviceName = Build.MANUFACTURER + ' ' + Build.MODEL,
                 customSharePhoneNumberE164 = phoneNumber.e164Formatted(),
-                phoneNumberTimestamp = timestamp,
+                phoneNumberTimestamp = epochUnixTimeSeconds,
                 socialShareOwnerEmail = socialShareOwnerId,
-                emailTimestamp = timestamp,
+                emailTimestamp = epochUnixTimeSeconds,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/gateway/repository/mapper/GatewayServiceCreateWalletMapper.kt:133]: emailTimestampSec 
metaTimestampSec
for future reference

File: app/src/main/java/org/p2p/wallet/auth/gateway/repository/mapper/GatewayServiceErrorMapper.kt [hunk 5]
```diff
@@ -48,4 +49,47 @@ class GatewayServiceErrorMapper {
             }
         }
             .also { Timber.tag("GatewayServiceMapper").i(error.toString()) }
+
+    fun fromNetwork(error: RpcResponse.Error): PushServiceError =
+        when (error.code) {
+            -32050 -> PushServiceError.TemporaryFailure(
+                error.code, error.message
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/gateway/repository/mapper/GatewayServiceErrorMapper.kt:56]: named arg

File: app/src/main/java/org/p2p/wallet/auth/gateway/repository/mapper/GatewayServiceUpdateMetadataMapper.kt [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    private class UpdateMetadataStruct(
+        val ethereumAddress: String,
+        val solanaPublicKey: String,
+        val epochUnixTime: Long,
+        val encryptedMetadata: String,
+    ) : BorshSerializable {
+        override fun serializeSelf(): ByteArray {
+            return getBorshBuffer()
+                .write(ethereumAddress, solanaPublicKey, epochUnixTime, encryptedMetadata)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/gateway/repository/mapper/GatewayServiceUpdateMetadataMapper.kt:33]: I can't see the 4 field in the specification. Request to update it

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/auth/gateway/api/request/UpdateMetadataRpcRequest.kt (hunk 1 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/auth/gateway/api/request/UpdateMetadataRpcRequest.kt:14]: it is json.toString

File: app/src/main/java/org/p2p/wallet/auth/gateway/api/response/UpdateMetadataResponse.kt (hunk 2 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/auth/gateway/api/response/UpdateMetadataResponse.kt:7]: https://p2pvalidator.atlassian.net/wiki/spaces/Wallet/pages/2479030273/API+Gateway+Get+metadata#Output-parameters.1

File: app/src/main/java/org/p2p/wallet/auth/gateway/repository/mapper/GatewayServiceUpdateMetadataMapper.kt (hunk 6 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/auth/gateway/repository/mapper/GatewayServiceUpdateMetadataMapper.kt:33]: https://p2pvalidator.atlassian.net/wiki/spaces/Wallet/pages/2479030273/API+Gateway+Get+metadata#Input-parameters.1

struct GetMetadataMessage<'a> {
    ethereum_id: &'a str,
    solana_pubkey: &'a str,
    timestamp_device: i64,
    metadata &'a [u8]
}

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/auth/gateway/repository/mapper/GatewayServiceErrorMapper.kt [hunk 7]
```diff
@@ -48,4 +49,49 @@ class GatewayServiceErrorMapper {
             }
         }
             .also { Timber.tag("GatewayServiceMapper").i(error.toString()) }
+
+    fun fromNetwork(error: RpcResponse.Error): PushServiceError =
+        when (error.code) {
+            -32050 -> PushServiceError.TemporaryFailure(
```
Comment (eduardmaximovich) [inline app/src/main/java/org/p2p/wallet/auth/gateway/repository/mapper/GatewayServiceErrorMapper.kt:55]: what do you think to write it in more declarative way?

something like this:
enum class PushServiceErrorCode(val clazz: KClass<Any>, val supportedCodes: Set<Int>) {
   TemporaryFailure(PushServiceError.TemporaryFailure::class, setOf(-32050)),
   PhoneNumberAlreadyConfirmed(PushServiceError.PhoneNumberAlreadyConfirmed::class, setOf(-32051));

   companion {
      fun createByCode(code: Int, message: String): PushServiceError {
         val error = values.filter { code in it.supportedCodes }.firstOrNull()
         return error.clazz.primaryConstructor?.call(code, message) ?: PushServiceError.UnknownFailure(code, message)
      }
   }
}

File: app/src/test/java/org/p2p/wallet/auth/gateway/repository/mapper/GatewayServiceOnboardingMetadataCipherTest.kt [hunk 8]
```diff
@@ -64,4 +69,38 @@ class GatewayServiceOnboardingMetadataCipherTest {
         // then
         assertThat(actualGatewayMetadata).isEqualTo(givenGatewayMetadata)
     }
+
+    @Test
```
Comment (eduardmaximovich) [inline app/src/test/java/org/p2p/wallet/auth/gateway/repository/mapper/GatewayServiceOnboardingMetadataCipherTest.kt:73]: you're my hero ❤️

**Review by eduardmaximovich — APPROVED** [review]

**Pull request merged**